bin/yk-auth

```diff
@@ -28,13 +28,50 @@ if [ -n "$pass_hash" ]; then
+if [ -r "$CONFIG_DIR/vm" ]; then
+	vm=$(cat $CONFIG_DIR/vm | grep -v '^#' | tr -d '\n')
+else
+	vm="sys-usb"
+fi
+
+# to use qubes-yubikey-dom0 in a configuration compatible with a NitroKey3, execute the following in dom0:
+# echo -n "your-20-digit-secret" | base32 | sudo tee /etc/qubes/yk-keys/nk-hotp-secret
+[ -r "$CONFIG_DIR/nk-hotp-secret" ] && secret=$(cat $CONFIG_DIR/nk-hotp-secret | grep -v '^#' | tr -d '\n')
 
-correct_response=$(echo $challenge | xxd -r -ps | openssl dgst -sha1 -macopt hexkey:$key -mac HMAC -r | cut -f1 -d ' ')
-# A yubikey configured for a variable length challenge will truncate the last byte when given a 64 byte challenge
-# The 'correct' response for a yubikey in that configuration will be:
-alternate_response=$(echo $challenge | cut -c1-126 | xxd -r -ps | openssl dgst -sha1 -macopt hexkey:$key -mac HMAC -r | cut -f1 -d ' ')
+if [ -z "$secret" ]; then	# YubiKey
+	if [ -r "$CONFIG_DIR/yk-slot" -a -r "$CONFIG_DIR/yk-secret-key.hex" ]; then
+		ykslot=$(cat $CONFIG_DIR/yk-slot | grep -v '^#' | tr -d '\n')
+		key=$(cat $CONFIG_DIR/yk-secret-key.hex | grep -v '^#' | tr -d ' \n')
+		challenge=$(head -c64 /dev/urandom | xxd -c 64 -ps)
+		response=$(qvm-run -a -u root --nogui -p $vm "ykchalresp -'$ykslot' -x '$challenge'")
+		correct_response=$(echo $challenge | xxd -r -ps | openssl dgst -sha1 -macopt hexkey:$key -mac HMAC -r | cut -f1 -d ' ')
+		# A yubikey configured for a variable length challenge will truncate the last byte when given a 64 byte challenge
+		# The 'correct' response for a yubikey in that configuration will be:
+		alternate_response=$(echo $challenge | cut -c1-126 | xxd -r -ps | openssl dgst -sha1 -macopt hexkey:$key -mac HMAC -r | cut -f1 -d ' ')
+
+		test "x$correct_response" = "x$response" -o "x$alternate_response" = "x$response"
+		exit $?
+	fi
+else				# NitroKey3
+	# put the output of the following command (replace the secret in quotes) into $CONFIG_DIR/nk-hotp-secret and use as secret for your NK3 as well
+	#secret=$(echo -n "your-20-digit-secret" | base32)	
+	# for registering that secret with your NK3 use (in the VM it's attached to):
+	#nitropy nk3 secrets register --kind hotp --hash sha256 --digits-str 8 --counter-start 1 --touch-button loginxs $secret
+	if [ -r "$CONFIG_DIR/nk-hotp-counter" ]; then
+		counter=$(cat $CONFIG_DIR/nk-hotp-counter | grep -v '^#' | tr -d '\n')
+	else
+		counter=1
+	fi
+	# HOTP implementation based on MIT-licensed code at github.com/bfreis/otp
+	key=$(echo -n "${secret}" | tr a-z A-Z | sed -e 's/ //g' | base32 -d | hexdump -ve '/1 "%02X"')
+	hmac=$(printf "$(printf "%016X" "$counter" | sed -e 's/\(..\)/\\x\1/g')" | 
+		openssl dgst -sha256 -mac HMAC -macopt "hexkey:$key" | sed -e 's/^.* //')
+	otp=$(printf "%08d" "$(( ( (0x${hmac:$((2 * 0x${hmac: -1})):8}) & 0x7FFFFFFF) % (10**8) ))" )
+	response=$(qvm-run -a --nogui -p $vm "\$HOME/.local/bin/nitropy nk3 secrets get loginxs" | tail -n1 | tr -d "\n")
```

Comment:
How is `nitropy` installed? I take there is no proper distribution package, right? Unfortunately, pip (or pipx for that matter) has rather poor integrity validation and is something we do not recommend using. I see there is also an option to download binary (binary? isn't that python?) directly from https://github.com/Nitrokey/pynitrokey/releases/tag/v0.4.44, but it isn't signed either.
Lets see, maybe this can be fixed: https://github.com/Nitrokey/pynitrokey/issues/495

Reply:
Yes, `pipx` seems to be the main way for the typical end user.

The "binary" is actually an executable Linux binary; unpacking it and doing `sudo ln -s /path/to/unpacked/binary /usr/local/bin/nitropy` allows easy execution without calling python3; it may still use python3 "under the hood", but who knows; when I uninstall python3 in a dvm (as far as possible; the process crashes because qubes packages need python3) and via `qvm-console-dispvm` unpack and execute the binary, it works.

Signed artifacts or even better official distro packaging would be nice, yes.